### PR TITLE
Add Pangoro2

### PR DESCRIPTION
**Pull request by AurevoirXavier**

### CODE REVIEW

File: node/Cargo.toml [hunk 1]
```diff
@@ -134,6 +135,7 @@ all-natives = [
 	"crab-native",
 	"darwinia-native",
 	"pangolin-native",
+	"pangoro-native",
```
Comment (boundless-forest) [inline node/Cargo.toml:138]: Add `"pangoro-runtime?/runtime-benchmarks"`  and `"pangoro-runtime?/try-runtime"`?

File: runtime/pangoro/src/pallets/deposit.rs [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Darwinia. If not, see <https://www.gnu.org/licenses/>.
+
+// darwinia
+use crate::*;
+
+pub enum PKtonMinting {}
```
Comment (boundless-forest) [inline runtime/pangoro/src/pallets/deposit.rs:22]: `PKtonMinting` -> `OKtonMinting`

File: runtime/pangoro/src/pallets/evm.rs [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+		Self::used_addresses().contains(&address)
+	}
+}
+
+impl pallet_evm::Config for Runtime {
+	type AddressMapping = FromH160;
+	type BlockGasLimit = BlockGasLimit;
+	type BlockHashMapping = pallet_ethereum::EthereumBlockHashMapping<Self>;
+	type CallOrigin = pallet_evm::EnsureAddressRoot<AccountId>;
+	type ChainId = ConstU64<43>;
```
Comment (boundless-forest) [inline runtime/pangoro/src/pallets/evm.rs:177]: `43` -> `45`

File: runtime/pangoro/tests/mock.rs [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+	pub fn build(&mut self) -> TestExternalities {
+		let mut t = frame_system::GenesisConfig::default().build_storage::<Runtime>().unwrap();
+
+		pallet_balances::GenesisConfig::<Runtime> { balances: self.balances.clone() }
+			.assimilate_storage(&mut t)
+			.unwrap();
+		pallet_assets::GenesisConfig::<Runtime> {
+			assets: vec![(KTON_ID, ROOT, true, 1)],
+			metadata: vec![(KTON_ID, b"Pangolin Commitment Token".to_vec(), b"OKTON".to_vec(), 18)],
```
Comment (boundless-forest) [inline runtime/pangoro/tests/mock.rs:53]: `Pangolin` -> `Pangoro`

File: runtime/pangoro/src/pallets/staking.rs [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+//
+// You should have received a copy of the GNU General Public License
+// along with Darwinia. If not, see <https://www.gnu.org/licenses/>.
+
+// darwinia
+use crate::*;
+
+fast_runtime_or_not!(MinStakingDuration, ConstU32<{ 5 * MINUTES }>, ConstU32<{ 14 * DAYS }>);
+
+pub enum PRingStaking {}
```
Comment (jiguantong) [inline runtime/pangoro/src/pallets/staking.rs:24]: ORing

Comment (jiguantong) [inline runtime/pangoro/src/pallets/staking.rs:24]: ORing?

File: runtime/pangoro/src/pallets/staking.rs [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+	fn unstake(who: &Self::AccountId, item: Self::Item) -> sp_runtime::DispatchResult {
+		<Balances as frame_support::traits::Currency<_>>::transfer(
+			&darwinia_staking::account_id(),
+			who,
+			item,
+			frame_support::traits::ExistenceRequirement::AllowDeath,
+		)
+	}
+}
+pub enum PKtonStaking {}
```
Comment (jiguantong) [inline runtime/pangoro/src/pallets/staking.rs:47]: OKton

File: runtime/pangoro/src/pallets/evm.rs [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+
+pub struct AssetIdConverter;
+impl darwinia_precompile_assets::AccountToAssetId<AccountId, AssetId> for AssetIdConverter {
+	fn account_to_asset_id(account_id: AccountId) -> AssetId {
+		let addr: sp_core::H160 = account_id.into();
+		addr.to_low_u64_be()
+	}
+}
+
+pub struct PangolinPrecompiles<R>(sp_std::marker::PhantomData<R>);
```
Comment (jiguantong) [inline runtime/pangoro/src/pallets/evm.rs:83]: PangoroPrecompiles

**Review by jiguantong — DISMISSED** [review]

File: node/src/chain_spec/pangoro.rs [hunk 8]
```diff
@@ -0,0 +1,362 @@
...
+// substrate
+use sc_chain_spec::Properties;
+use sc_service::ChainType;
+use sp_core::{crypto::UncheckedInto, H160};
+
+/// Specialized `ChainSpec` for the normal parachain runtime.
+pub type ChainSpec = sc_service::GenericChainSpec<GenesisConfig, Extensions>;
+
+fn properties() -> Properties {
+	super::properties("ORING", 42)
```
Comment (boundless-forest) [inline node/src/chain_spec/pangoro.rs:43]: I guess this should be 18 instead of 42, the same as darwinia.

File: runtime/pangoro/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,597 @@
...
+//
+// Darwinia is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Darwinia. If not, see <https://www.gnu.org/licenses/>.
+
+//! Darwinia runtime.
```
Comment (boundless-forest) [inline runtime/pangoro/src/lib.rs:19]: `Pangoro` runtime

File: runtime/pangoro/src/migration.rs [hunk 10]
```diff
@@ -0,0 +1,25 @@
+// darwinia
```
Comment (boundless-forest) [inline runtime/pangoro/src/migration.rs:1]: Add License?

File: runtime/pangoro/src/pallets/system.rs [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+	/// Converts a module to an index of this module in the runtime.
+	type PalletInfo = PalletInfo;
+	/// The aggregated dispatch type that is available for extrinsics.
+	type RuntimeCall = RuntimeCall;
+	/// The ubiquitous event type.
+	type RuntimeEvent = RuntimeEvent;
+	/// The ubiquitous origin type.
+	type RuntimeOrigin = RuntimeOrigin;
+	/// This is used as an identifier of the chain. 42 is the generic substrate prefix.
+	type SS58Prefix = ConstU16<18>;
```
Comment (boundless-forest) [inline runtime/pangoro/src/pallets/system.rs:104]: Checked out the config in the darwinia-1.0 repo. for the darwinia chain, SS58Prefix=18, for the crab chain, SS58Prefix=42.

But in the darwinia 2.0 repo, both of them are 18. Is this expected?

**Review by boundless-forest — COMMENTED** [review]
Some feedback left, other changes LGTM.

**Review by boundless-forest — APPROVED** [review]

**Pull request merged**